api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1760,162 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+		_, oldSubclusterInNewSandboxes := newSuclusterInSandbox[oldSubclusterName]
+		_, oldSubclusterInPersist := newSubclusterMap[oldSubclusterName]
+		// for unsandboxing, check shutdown field of the subcluster and sandbox
+		if !oldSubclusterInNewSandboxes && oldSubclusterInPersist {
+			oldSubcluster := oldSubclusterMap[oldSubclusterName]
+			oldSandbox := oldSandboxMap[oldSandboxName]
+			oldSubclusterIndex := oldSubclusterIndexMap[oldSubclusterName]
+			if oldSubcluster.Shutdown || v.Status.Subclusters[oldSubclusterIndex].Shutdown {
+				p := field.NewPath("spec").Child("subclusters")
+				err := field.Invalid(p,
+					oldSubclusterName,
+					fmt.Sprintf("cannot unsandbox subcluster %q that has Shutdown field set to true",
+						oldSubclusterName))
+				allErrs = append(allErrs, err)
+				continue
+			}
+			if oldSandbox.Shutdown {
+				p := field.NewPath("spec").Child("sandboxes")
+				err := field.Invalid(p,
+					oldSubclusterName,
+					fmt.Sprintf("cannot unsandbox subcluster %q in sandbox %q that has Shutdown field set to true",
+						oldSubclusterName, oldSandboxName))
+				allErrs = append(allErrs, err)
+				continue
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensure: when a sandbox is to be shutdown, if a subcluster in it has annotation vertica.com/shutdown-driven-by-sandbox set to true, the
+// subcluster's shutdown field should not be updated
+func (v *VerticaDB) validateAnnotatedSubclustersInShutdownSandbox(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	newSandboxMap := v.GenSandboxMap()
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSubclusterIndexMap := v.GenSubclusterIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterMap := oldObj.GenSubclusterMap()
+	for _, sandboxName := range persistSandboxes {
+		newSandbox := newSandboxMap[sandboxName]
+		oldSandbox := oldSandboxMap[sandboxName]
+		if newSandbox.Shutdown {
+			for _, subclusterName := range oldSandbox.Subclusters {
+				oldSubcluster := oldSubclusterMap[subclusterName.Name]
+				if drivenby, ok := oldSubcluster.Annotations["vertica.com/shutdown-driven-by-sandbox"]; ok && drivenby == trueString {
+					newSubcluster, oldSclusterPersist := newSubclusterMap[subclusterName.Name]
+					if oldSclusterPersist && oldSubcluster.Shutdown != newSubcluster.Shutdown {
+						index := newSubclusterIndexMap[subclusterName.Name]
+						p := field.NewPath("spec").Child("subclusters")
+						err := field.Invalid(p.Index(index),
+							subclusterName.Name,
+							fmt.Sprintf("cannot change Shutdown field for subcluster %q that has annotation \"vertica.com/shutdown-driven-by-sandbox\" set to true",
+								subclusterName.Name))
+						allErrs = append(allErrs, err)
+					}
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+func (v *VerticaDB) validateShutdownSandboxImage(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
```

Comment:
We should have one more rule: if the sandbox is shut down or partial shut down, we cannot add or remove any subclusters in it.

Reply:
The remove scenario is covered by unsandbox. We can discuss with Roy about the add scenario.

Reply:
FYI:
I did some manual testing. When I tried to add a new subcluster to a shutdown sandbox. Webhook did not complain. The spec was updated but nothing happened. It seems that is not supported and we may need a new rule.

Reply:
The proposed new rule can be found in another webhook JIRA VER-97254